qemu_riscv64: Failure in CMakeFiles/app.dir/src/zephyr

**Describe the bug**
On qemu_riscv64 platform, CMakeFiles/app.dir/src/zephyr/arch/riscv/core/arch_util.c.obj and CMakeFiles/app.dir/src/zephyr/timer/bench_riscv_machine_timer.c.obj failed.

Please also mention any information which could help others to understand
the problem you're facing:
 - qemu_riscv64
 - error: 'RISCV_MACHINE_TIMER_IRQ' undeclared (first use in this function)
 - error: 'RISCV_MTIME_BASE' undeclared (first use in this function)

**To Reproduce**
Steps to reproduce the behavior:
1. cmake -GNinja -DRTOS=zephyr -DBOARD=qemu_riscv64 -S . -B build
2. cd build
3. ninja run

**Logs and console output**
```
FAILED: CMakeFiles/app.dir/src/zephyr/arch/riscv/core/arch_util.c.obj
ccache /home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I/home/zhao/rtos-benchmark/h -I/home/zhao/zephyrproject/zephyr/include/zephyr -I/home/zhao/zephyrproject/zephyr/include -I/home/zhao/rtos-benchmark/build2/zephyr/include/generated -I/home/zhao/zephyrproject/zephyr/soc/riscv/riscv-privilege/virt -I/home/zhao/zephyrproject/zephyr/soc/riscv/riscv-privilege/common/. -I/home/zhao/zephyrproject/zephyr/subsys/testsuite/include -I/home/zhao/zephyrproject/zephyr/subsys/testsuite/include/zephyr -isystem /home/zhao/zephyrproject/zephyr/lib/libc/minimal/include -isystem /home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/include -isystem /home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/include-fixed -DITERATIONS=10000 -DCALIBRATION_LOOPS=10000 -DZEPHYR -Os -imacros /home/zhao/rtos-benchmark/build2/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcmodel=medany -mabi=lp64 -march=rv64imac --sysroot=/home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/riscv64-zephyr-elf -imacros /home/zhao/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/zhao/rtos-benchmark=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/zhao/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/zhao/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/zephyr/arch/riscv/core/arch_util.c.obj -MF CMakeFiles/app.dir/src/zephyr/arch/riscv/core/arch_util.c.obj.d -o CMakeFiles/app.dir/src/zephyr/arch/riscv/core/arch_util.c.obj -c /home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c
/home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c: In function 'bench_timer_isr_get':
/home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c:15:44: error: 'RISCV_MACHINE_TIMER_IRQ' undeclared (first use in this function); did you mean 'RISCV_MACHINE_EXT_IRQ'?
   15 |  return (bench_isr_handler_t)_sw_isr_table[RISCV_MACHINE_TIMER_IRQ].isr;
      |                                            ^~~~~~~~~~~~~~~~~~~~~~~
      |                                            RISCV_MACHINE_EXT_IRQ
/home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c:15:44: note: each undeclared identifier is reported only once for each function it appears in
/home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c: In function 'bench_timer_isr_set':
/home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c:36:16: error: 'RISCV_MACHINE_TIMER_IRQ' undeclared (first use in this function); did you mean 'RISCV_MACHINE_EXT_IRQ'?
   36 |  _sw_isr_table[RISCV_MACHINE_TIMER_IRQ].isr = (void(*)(const void *))isr;
      |                ^~~~~~~~~~~~~~~~~~~~~~~
      |                RISCV_MACHINE_EXT_IRQ
/home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c: In function 'bench_timer_isr_get':
/home/zhao/rtos-benchmark/src/zephyr/arch/riscv/core/arch_util.c:16:1: warning: control reaches end of non-void function [-Wreturn-type]
   16 | }
      | ^
```
```
[11/149] Building C object CMakeFiles/app.dir/src/zephyr/timer/bench_riscv_machine_timer.c.obj
FAILED: CMakeFiles/app.dir/src/zephyr/timer/bench_riscv_machine_timer.c.obj
ccache /home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I/home/zhao/rtos-benchmark/h -I/home/zhao/zephyrproject/zephyr/include/zephyr -I/home/zhao/zephyrproject/zephyr/include -I/home/zhao/rtos-benchmark/build2/zephyr/include/generated -I/home/zhao/zephyrproject/zephyr/soc/riscv/riscv-privilege/virt -I/home/zhao/zephyrproject/zephyr/soc/riscv/riscv-privilege/common/. -I/home/zhao/zephyrproject/zephyr/subsys/testsuite/include -I/home/zhao/zephyrproject/zephyr/subsys/testsuite/include/zephyr -isystem /home/zhao/zephyrproject/zephyr/lib/libc/minimal/include -isystem /home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/include -isystem /home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/include-fixed -DITERATIONS=10000 -DCALIBRATION_LOOPS=10000 -DZEPHYR -Os -imacros /home/zhao/rtos-benchmark/build2/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcmodel=medany -mabi=lp64 -march=rv64imac --sysroot=/home/zhao/zephyr-sdk-0.14.2/riscv64-zephyr-elf/riscv64-zephyr-elf -imacros /home/zhao/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/zhao/rtos-benchmark=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/zhao/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/zhao/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/zephyr/timer/bench_riscv_machine_timer.c.obj -MF CMakeFiles/app.dir/src/zephyr/timer/bench_riscv_machine_timer.c.obj.d -o CMakeFiles/app.dir/src/zephyr/timer/bench_riscv_machine_timer.c.obj -c /home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c: In function 'mtime':
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c:18:38: error: 'RISCV_MTIME_BASE' undeclared (first use in this function); did you mean 'RISCV_MSIP_BASE'?
   18 |         return *(volatile uint64_t *)RISCV_MTIME_BASE;
      |                                      ^~~~~~~~~~~~~~~~
      |                                      RISCV_MSIP_BASE
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c:18:38: note: each undeclared identifier is reported only once for each function it appears in
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c: In function 'get_hart_mtimecmp':
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c:35:9: error: 'RISCV_MTIMECMP_BASE' undeclared (first use in this function); did you mean 'RISCV_MSIP_BASE'?
   35 |  return RISCV_MTIMECMP_BASE + (_current_cpu->id * 8);
      |         ^~~~~~~~~~~~~~~~~~~
      |         RISCV_MSIP_BASE
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c: In function 'mtime':
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c:31:1: warning: control reaches end of non-void function [-Wreturn-type]
   31 | }
      | ^
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c: In function 'get_hart_mtimecmp':
/home/zhao/rtos-benchmark/src/zephyr/timer/bench_riscv_machine_timer.c:36:1: warning: control reaches end of non-void function [-Wreturn-type]
   36 | }
      | ^
[19/149] Building C object CMakeFiles/app.dir/src/zephyr/bench_porting_layer_zephyr.c.obj
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain - SDK 14.1

--- a/src/zephyr/arch/arm/cortex_m/arch_util.c
+++ b/src/zephyr/arch/arm/cortex_m/arch_util.c
@@ -1,4 +1,4 @@
-#include <arch/arm/aarch32/cortex_m/cmsis.h>
+#include <zephyr/arch/arm/aarch32/cortex_m/cmsis.h>
 
 #include "bench_api.h"
 #include "bench_utils.h"

--- a/src/zephyr/arch/riscv/core/arch_util.c
+++ b/src/zephyr/arch/riscv/core/arch_util.c
@@ -2,17 +2,18 @@
 #include "bench_utils.h"
 
 extern struct _isr_table_entry _sw_isr_table[];
+extern uint32_t riscv_machine_timer_irq;
 
 bench_isr_handler_t bench_timer_isr_get(void)
 {
 
 	/*
 	 * At the current time, it is assumed that the timer interrupt
-	 * is NOT a level 2 interrupt and RISCV_MACHINE_TIMER_IRQ can
+	 * is NOT a level 2 interrupt and riscv_machine_timer_irq can
 	 * simply be an index into _sw_isr_table[].
 	 */
 
-	return (bench_isr_handler_t)_sw_isr_table[RISCV_MACHINE_TIMER_IRQ].isr;
+	return (bench_isr_handler_t)_sw_isr_table[riscv_machine_timer_irq].isr;
 }
 
 uint32_t bench_timer_cycles_per_second(void)
@@ -33,7 +34,7 @@ uint32_t bench_timer_cycles_per_tick(void)
 
 void bench_timer_isr_set(bench_isr_handler_t  isr)
 {
-	_sw_isr_table[RISCV_MACHINE_TIMER_IRQ].isr = (void(*)(const void *))isr;
+	_sw_isr_table[riscv_machine_timer_irq].isr = (void(*)(const void *))isr;
 
 	return;
 }

--- a/src/zephyr/bench_porting_layer_zephyr.c
+++ b/src/zephyr/bench_porting_layer_zephyr.c
@@ -2,9 +2,9 @@
 
 #include "bench_api.h"
 #include "bench_porting_layer_zephyr.h"
-#include <zephyr.h>
-#include <timing/timing.h>
-#include <irq_offload.h>
+#include <zephyr/zephyr.h>
+#include <zephyr/timing/timing.h>
+#include <zephyr/irq_offload.h>
 
 /*
  * Constants.

--- a/src/zephyr/bench_porting_layer_zephyr.h
+++ b/src/zephyr/bench_porting_layer_zephyr.h
@@ -3,9 +3,9 @@
 #ifndef PORTING_LAYER_ZEPHYR_H_
 #define PORTING_LAYER_ZEPHYR_H_
 
-#include <zephyr.h>
-#include <kernel.h>
-#include <timing/timing.h>
+#include <zephyr/zephyr.h>
+#include <zephyr/kernel.h>
+#include <zephyr/timing/timing.h>
 #include <stdio.h>
 
 typedef timing_t bench_time_t;

--- a/src/zephyr/timer/bench_cortex_m_systick.c
+++ b/src/zephyr/timer/bench_cortex_m_systick.c
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <arch/arm/aarch32/cortex_m/cmsis.h>
+#include <zephyr/arch/arm/aarch32/cortex_m/cmsis.h>
 
 #include "bench_api.h"
 #include "bench_utils.h"

--- a/src/zephyr/timer/bench_riscv_machine_timer.c
+++ b/src/zephyr/timer/bench_riscv_machine_timer.c
@@ -9,15 +9,54 @@
 
 #define MIN_DELAY  1000ULL
 
+/* neorv32-machine-timer */
+#if DT_HAS_COMPAT_STATUS_OKAY(andestech_machine_timer)
+#define DT_DRV_COMPAT andestech_machine_timer
+
+#define MTIME_REG	DT_INST_REG_ADDR(0)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 8)
+#define TIMER_IRQN	DT_INST_IRQN(0)
+/* neorv32-machine-timer */
+#elif DT_HAS_COMPAT_STATUS_OKAY(neorv32_machine_timer)
+#define DT_DRV_COMPAT neorv32_machine_timer
+
+#define MTIME_REG	DT_INST_REG_ADDR(0)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 8)
+#define TIMER_IRQN	DT_INST_IRQN(0)
+/* nuclei,systimer */
+#elif DT_HAS_COMPAT_STATUS_OKAY(nuclei_systimer)
+#define DT_DRV_COMPAT nuclei_systimer
+
+#define MTIME_REG	DT_INST_REG_ADDR(0)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 8)
+#define TIMER_IRQN	DT_INST_IRQN(0)
+/* sifive,clint0 */
+#elif DT_HAS_COMPAT_STATUS_OKAY(sifive_clint0)
+#define DT_DRV_COMPAT sifive_clint0
+
+#define MTIME_REG	(DT_INST_REG_ADDR(0) + 0xbff8U)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 0x4000U)
+#define TIMER_IRQN	DT_INST_IRQ_BY_IDX(0, 1, irq)
+/* telink,machine-timer */
+#elif DT_HAS_COMPAT_STATUS_OKAY(telink_machine_timer)
+#define DT_DRV_COMPAT telink_machine_timer
+
+#define MTIME_REG	DT_INST_REG_ADDR(0)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 8)
+#define TIMER_IRQN	DT_INST_IRQN(0)
+#endif
+
+uint32_t riscv_machine_timer_irq = TIMER_IRQN;
+
 /**
  * @brief Read the current cycle count from the MTIME register
  */
 static uint64_t mtime(void)
 {
 #ifdef CONFIG_64BIT
-        return *(volatile uint64_t *)RISCV_MTIME_BASE;
+        return *(volatile uint64_t *)MTIME_REG;
 #else
-        volatile uint32_t *r = (uint32_t *)RISCV_MTIME_BASE;
+        volatile uint32_t *r = (uint32_t *)MTIME_REG;
         uint32_t  lo;
         uint32_t  hi;
 
@@ -32,7 +71,7 @@ static uint64_t mtime(void)
 
 static uint64_t get_hart_mtimecmp(void)
 {
-	return RISCV_MTIMECMP_BASE + (_current_cpu->id * 8);
+	return MTIMECMP_REG + (_current_cpu->id * 8);
 }
 
 /**

